### PR TITLE
Organizations and locations

### PR DESCRIPTION
**Pull request by abenari**
First cut

### CODE REVIEW

**Comment by ohadlevy** [comment]
@abenari please rebase, as its not mergable

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -9,8 +9,7 @@ class ApplicationController < ActionController::Base
   # standard layout to all controllers
   helper 'layout'
 
-  before_filter :require_ssl, :require_login
-  before_filter :require_mail
```
Comment (ohadlevy) [inline app/controllers/application_controller.rb]: that line is missing from the merge i guess?

File: app/helpers/taxonomy_helper.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+module TaxonomyHelper
+  def show_location_tab?
+    SETTINGS[:locations_enabled] #&& (User.current.admin? || authorized_keys(:locations, :index))
```
Comment (ohadlevy) [inline app/helpers/taxonomy_helper.rb]: commented out?

File: app/helpers/taxonomy_helper.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  def show_organization_tab?
+    SETTINGS[:organizations_enabled] #&& (User.current.admin? || authorized_keys(:organizations, :index))
+  end
+
+  def show_taxonomy_tabs?
+    SETTINGS[:locations_enabled] or SETTINGS[:organizations_enabled]
+  end
+
+  def show_add_location_button?
+    Location.all.count == 0 #and (User.current.admin? || authorized_keys(:locations, :add))
```
Comment (ohadlevy) [inline app/helpers/taxonomy_helper.rb]: .all is not required, just Location.count

File: app/models/domain.rb [hunk 4]
```diff
@@ -20,12 +21,22 @@ class Domain < ActiveRecord::Base
...
+    with_taxonomy_scope do
+      order("LOWER(domains.name)")
+    end
+  }
 
   class Jail < Safemode::Jail
     allow :name, :fullname
   end
 
+  def taxonomies
```
Comment (ohadlevy) [inline app/models/domain.rb]: what does this do?

File: app/models/hostgroup.rb [hunk 5]
```diff
@@ -52,6 +57,7 @@ class Jail < Safemode::Jail
   end
 
   #TODO: add a method that returns the valid os for a hostgroup
+  #
```
Comment (ohadlevy) [inline app/models/hostgroup.rb]: whitespace

File: app/models/subnet.rb [hunk 6]
```diff
@@ -18,6 +19,8 @@ class Subnet < ActiveRecord::Base
   default_scope :order => 'priority'
   validate :validate_ranges
 
+  #default_scope lambda { with_taxonomy_scope }
```
Comment (ohadlevy) [inline app/models/subnet.rb]: why commented out?

File: app/models/taxable_taxonomy.rb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+class TaxableTaxonomy < ActiveRecord::Base
+  belongs_to :taxonomy
```
Comment (ohadlevy) [inline app/models/taxable_taxonomy.rb:2]: should we audit this?

File: app/models/taxonomix.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      @primary_key = "taxable_id"
+      has_many @taxonomy_join_table, :dependent => :destroy, :as => :taxable
+      has_many :locations, :through => @taxonomy_join_table,     :source => :taxonomy
+      has_many :organizations, :through => @taxonomy_join_table, :source => :taxonomy
+      after_initialize :set_current_taxonomy
+
+      def self.with_taxonomy_scope
+        scope =  block_given? ? yield : where(1)
+
+        scope = scope.joins(" INNER JOIN #{@taxonomy_join_table} loc1 ON loc1.#{@primary_key} = #{self.table_name}.id").where("loc1.taxonomy_id in (?)", Location.current.id.to_a) if SETTINGS[:locations_enabled] and Location.current
```
Comment (ohadlevy) [inline app/models/taxonomix.rb]: mind adding a bit of docs why we are doing this join this way?

File: app/models/user.rb [hunk 9]
```diff
@@ -62,6 +66,9 @@ def to_param
     "#{id}-#{login.parameterize}"
   end
 
+  def taxonomies
```
Comment (ohadlevy) [inline app/models/user.rb]: left over?

File: app/views/domains/_form.html.erb [hunk 10]
```diff
@@ -10,7 +16,7 @@
     <div class="tab-pane active" id="primary">
       <%= text_f   f, :name, :help_inline => "The full DNS Domain name" %>
       <%= text_f   f, :fullname, :help_inline => "Full name describing the domain" %>
-      <%= select_f f, :dns_id, Feature.find_by_name("DNS").smart_proxies, :id, :name,
+      <%# select_f f, :dns_id, Feature.find_by_name("DNS").smart_proxies, :id, :name,
```
Comment (ohadlevy) [inline app/views/domains/_form.html.erb]: ?

File: app/views/home/_location_dropdown.rhtml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+<li class="dropdown">
+  <%= link_to ((Location.current.name rescue("Locations")) + content_tag(:span,'', :class=>"caret")).html_safe, "#", :class => "dropdown-toggle", :'data-toggle'=>"dropdown", :title => 'Current Location' %>
+  <ul class="dropdown-menu">
+    <% if show_add_location_button? %>
+      <div class="no-taxonomies">
+        <div class="alert">
+          <button type="button" class="close" data-dismiss="alert">×</button>
```
Comment (ohadlevy) [inline app/views/home/_location_dropdown.rhtml]: &times. ?

File: config/settings.yaml.example [hunk 12]
```diff
@@ -1,4 +1,12 @@
---- 
+---
+# You must restart the Rack server after changing either of these values
+# because they are cached at startup.
+#
+#your default puppet server - can be overridden in the host level
```
Comment (ohadlevy) [inline config/settings.yaml.example]: bad merge for the puppte_server value

File: app/models/location.rb [hunk 13]
```diff
@@ -0,0 +1,28 @@
+class Location < Taxonomy
+  include Foreman::ThreadSession::LocationModel
+
```
Comment (ohadlevy) [inline app/models/location.rb:3]: should we location/orgs to trends as well?

File: test/unit/user_test.rb [hunk 14]
```diff
@@ -71,12 +71,13 @@ def setup
...
-    @user.firstname = "Ali Al"
-    @user.lastname = "Salame"
+  test "to_label method should return a firstname and the lastname and username" do
+    firstname = @user.firstname = "Ali Al"
+    lastname = @user.lastname = "Salame"
+    login = @user.login = "foo"
     assert @user.save
 
-    assert_equal "Ali Al Salame", @user.to_label
+    assert_equal "#{firstname} #{lastname} (#{login})", @user.to_label
```
Comment (ohadlevy) [inline test/unit/user_test.rb]: whats the point of this change?

File: app/controllers/application_controller.rb [hunk 15]
```diff
@@ -290,4 +289,26 @@ def generic_exception(exception)
     render :template => "common/500", :layout => !request.xhr?, :status => 500, :locals => { :exception => exception}
   end
 
+  def set_taxonomy
+    return if User.current.nil?
+    orgs = Organization.my_organizations
+
+    Organization.current = if orgs.count == 1
```
Comment (ohadlevy) [inline app/controllers/application_controller.rb]: should we check if orgs are enabled first?

Comment (ohadlevy) [inline app/controllers/application_controller.rb]: or not run this method at all if its disabled?

File: app/controllers/locations_controller.rb [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+  def select
+    @location = params[:id] ? Location.find(params[:id]) : nil
+    Location.current = @location
+    session[:location_id] = @location ? @location.id : nil
+    redirect_back_or_to root_url
+  end
+
+  def load_vars_for_ajax
+    return unless @location
```
Comment (ohadlevy) [inline app/controllers/locations_controller.rb]: is this required?

File: app/models/user.rb [hunk 17]
```diff
@@ -54,7 +57,7 @@ class User < ActiveRecord::Base
   scoped_search :in => :roles, :on => :name, :rename => :role, :complete_value => true
 
   def to_label
-    "#{firstname} #{lastname}"
+    "#{firstname} #{lastname} (#{login})"
```
Comment (ohadlevy) [inline app/models/user.rb]: is this a left over?

File: app/views/users/_form.html.erb [hunk 18]
```diff
@@ -1,42 +1,67 @@
 <%= form_for @user, :html=>{:class=>"form-horizontal"} do |f| %>
-  <%= base_errors_for @user %>
+    <%= base_errors_for @user %>
```
Comment (ohadlevy) [inline app/views/users/_form.html.erb]: whitespace

File: test/unit/domain_test.rb [hunk 19]
```diff
@@ -49,6 +49,22 @@ def setup
...
+    location2 = Location.create :name => "Switzerland"
+    assert location2.save!
+
+    domain = Domain.create :name => "test.net"
+    domain.locations = []
+    domain.locations.push location1
+    domain.locations.push location2
+    assert domain.save!
+
+    puts domain.inspect
```
Comment (ohadlevy) [inline test/unit/domain_test.rb]: debug!

**Comment by ohadlevy** [comment]
merged!! thanks @skottler @abenari and @blomquisg

**Comment by skottler** [comment]
Woohoo!

**Pull request closed without merging**